validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,162 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+func (c beaconApiValidatorClient) waitForActivation(ctx context.Context, in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	// Represents the set of keys in `in`
+	targetPubKeys := make(map[string]struct{})
+
+	// Represents the set of keys actually returned by the beacon node
+	retrievedPubKeys := make(map[string]struct{})
+
+	// Contains all keys in targetPubKeys but not in retrievedPubKeys
+	missingPubKeys := make(map[string]struct{})
+
+	statuses := []*ethpb.ValidatorActivationResponse_Status{}
+
+	for _, publicKey := range in.PublicKeys {
+		stringPubKey := "0x" + hex.EncodeToString(publicKey)
+		targetPubKeys[stringPubKey] = struct{}{}
+	}
+
+	params := url.Values{}
+
+	for _, publicKey := range in.PublicKeys {
+		params.Add("id", fmt.Sprintf("0x%s", hex.EncodeToString(publicKey)))
+	}
+
+	resp, err := c.httpClient.Get(c.url + "/eth/v1/beacon/states/head/validators?" + params.Encode())
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err = resp.Body.Close(); err != nil {
+			return
+		}
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		errorJson := apimiddleware.DefaultErrorJson{}
+		err = json.NewDecoder(resp.Body).Decode(&errorJson)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, errors.Errorf("error %d: %s", errorJson.Code, errorJson.Message)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+
+	var res ValidatorsHttpResponse
+	if err := json.Unmarshal(body, &res); err != nil {
+		return nil, err
```

Comment:
Consider wrapping this error as well. It can really help with debugging.

```suggestion
		return nil, errors.Wrap(err, "failed to decode validators http response")
```

Reply:
Should I wrap all occurrences of this kind of code
```go
if err != nil {
    return nil, err
}
```

with `errors.Wrap` and?